Add steps for Infinite Tracing data sources for browser, mobile, and lambda (#5228)
newrelic/docs-website: src/content/docs/distributed-tracing/infinite-tracing/set-trace-observer.mdx

  - /docs/understand-dependencies/distributed-tracing/infinite-tracing/set-trace-observer
---

If you are following the [steps](/docs/distributed-tracing/enable-configure/quick-start) to set up distributed tracing and are continuing with the steps for the advanced option called Infinite Tracing, you need to set up a trace observer. The trace observer is what allows you to go beyond the basic [head-based sampling](/docs/distributed-tracing/concepts/how-new-relic-distributed-tracing-works/#head-based) of standard distributed tracing. 

> The trace observer is a cloud-based New Relic tool that uses [tail-based sampling](/docs/distributed-tracing/concepts/how-new-relic-distributed-tracing-works/#tail-based) to decide what trace data to keep and send to New Relic. The trace observer is an AWS-based service that provides you with a low-latency and low-cost way to send your telemetry data to New Relic. The trace observer sends data via our Trace API, which is the entry point for all distributed trace data we ingest.

Reviewer: I think it might help to make it clear that the trace observer is what allows Infinite Tracing to work. As it is, the two concepts don't seem exactly aligned, so one might be left with the impression that the trace observer is something separate or not directly related to Infinite Tracing. If that makes sense. So maybe something like 'The trace observer is the tool that implements Infinite Tracing's tail-based sampling. The trace observer is...'

Author: @zuluecho9, I modified a sentence in the previous paragraph: The trace observer is a component of Infinite Tracing that allows you to go beyond the basic [head-based sampling](/docs/distributed-tracing/concepts/how-new-relic-distributed-tracing-works/#head-based) of standard distributed tracing.

          </Callout>
     </Collapser>
   </CollapserGroup>

### Initial trace observer configuration for browser, mobile, and Lambda [#initial-config]

> Once you have a trace observer, you need to specify which data sources (services, apps, or entities) should have their data sent to the Infinite Tracing trace observer. Before you select your data sources, familiarize yourself with these issues:

Reviewer 2: I thought "issues" seemed a little too close to "problems". What do you think of this? Does it feel awkward?

```suggestion
Once you have a trace observer, you need to specify which data sources (services, apps, or entities) should have their data sent to the Infinite Tracing trace observer. Before you select your data sources, familiarize yourself with the following considerations:
```

<CollapserGroup>
  <Collapser
    className="freq-link"
    id="timing"
    title="How long before I see my data in the New Relic UI?"
   >

>    As soon as you select your browser, mobile, and Lambda data sources, you should expect a delay before you see your results in the New Relic UI. This delay may be anywhere from one to two minutes. 

Reviewer 2: ```suggestion
   When you select your browser, mobile, and Lambda data sources, you should expect a delay of up to three minutes before you see your results in the New Relic UI for that trace observer. This allows the trace observer to wait long enough for these high-latency data sources to arrive before making a sampling decision . 
```

>    When you select your browser, mobile, and Lambda data sources, you should expect a delay of up to three minutes before you see your results in the New Relic UI for that trace observer. This allows the trace observer to wait long enough for these high-latency data sources to arrive before making a sampling decision . 

Reviewer: ```suggestion
   When you select your browser, mobile, and Lambda data sources, you should expect a delay of up to three minutes before you see your results in the New Relic UI for that trace observer. This allows the trace observer to wait long enough for these high-latency data sources to arrive before making a sampling decision. 
```

Author: @zuluecho9, nice catch!

   For example, if you re-run the [sample curl test](#sample-request) test after you add browser, mobile, or Lambda data sources, you should expect a delay before the data appears in the UI.
  </Collapser>
  <Collapser
    className="freq-link"
    id="multi-named-apps"
    title="What if my browser app is connected to an APM agent that has multiple names?"
  >

>     If your browser app is associated with an APM application that has multiple names, you can only select one of the names when you choose your data sources. If you try to select the same application that goes by another name, the UI will prompt you that you need to choose only one name.

Reviewer 2: ```suggestion
    If your browser app is associated with an [APM application that has multiple names](/docs/apm/agents/manage-apm-agents/app-naming/use-multiple-names-app/), you can only select one from the group of multi-named browser apps when you choose your data source. If you try to select the same application that goes by another name, the UI will prompt you that you need to choose only one browser app.
```

  </Collapser>
  <Collapser
    className="freq-link"
    id="lambda-sampling"
    title="How does Lambda sampling work?"
  >
  By default, when you set up Lambda layers, you automatically get head-based sampling. When you enable Infinite Tracing for Lambda, your head-based sampling data is additionally sampled by the trace observer. The main benefit to enabling Infinite Tracing for your Lambda functions is to ensure their spans are included with spans from other services tracked by Infinite Tracing. 
  </Collapser>
</CollapserGroup>
 

> The instructions below are about using the New Relic UI, but if you prefer to take a programmatic approach, you can specify your trace observer data sources using [GraphQL](/docs/apis/nerdgraph/examples/configure-infinite-tracing-graphql/#updating-the-data-sources). When you finish, return to this page and continue with step 4 (Optional configurations) below. 

Reviewer: ```suggestion
The procedure below if for using the New Relic UI, but if you prefer a programmatic approach, you can [specify data sources with GraphQL](/docs/apis/nerdgraph/examples/configure-infinite-tracing-graphql/#updating-the-data-sources). If you use GraphQL, return to this procedure and continue with step 4 (optional configurations) below. 
```

Author: @zuluecho9, great suggestion!

To finish the configuration for browser, mobile, and Lambda in the UI: 

> 1. Return to the trace observer Nerdlet (**[one.newrelic.com](https://one.newrelic.com) > Apps > Your Apps > New Relic Edge**).

Reviewer: I would avoid using Nerdlet in this way, as it's generally just used in the context of creating New Relic One custom apps. I'd instead just refer to the 'New Relic Edge app.'.

Author: @zuluecho9, I went with "trace observer app."

2. For your trace observer, go to the end of the row, click the three dots, and then click **Edit mobile, browser, lambda sources**.

  ![Screenshot showing menu when you click the three dots next to your trace observer](./images/browser-mobile-lambda-menu.png "Screenshot showing menu when you click the three dots next to your trace observer")

3. Select the data sources you want to include, keeping the following in mind:

    * You can use the filter to narrow your options.

>     * If your browser app is connected to an APM agent has multiple names, you can can only choose one name.

Reviewer: ```suggestion
    * If your browser app is connected to an APM agent that has multiple names, you can can only choose one name.
```

Author: @zuluecho9 nice catch!

    * If an entry is gray, it is already being used by another trace observer.

  ![Screenshot showing the list of data sources you want to send tracing data](./images/select-data-sources.png "Screenshot showing the list of data sources you want to send tracing data")
  
4. Bookmark the [optional configurations](#follow-up-config) below so you can return to them after you experiment with Infinite Tracing.

> 5. View your Infinite Tracing data in the [New Relic UI](/docs/distributed-tracing/ui-data/understand-use-distributed-tracing-ui/).

Reviewer: a small thing, but I notice some of your link text isn't that descriptive. for this example, I would go something like: 

View your data in the [distributed tracing UI], 
instead of '[New Relic UI]. 

I think this is keeping with our general guidelines to try to make the link text slightly descriptive of what's there. In other words, [New Relic UI] I might suppose would take me to general info about New Relic One UI or something. I only mention cause I saw a couple other link texts that were pretty vague like that.

Author: @zuluecho9, good tip! I fixed this one and some other generally worded links!

6. If you have some services instrumented with New Relic APM agents or third-party telemetry tools, complete the steps in that [section](#apm-and-third-party) to ensure their data also gets to New Relic.

## (Optional) Additional configuration for all service types [#follow-up-config]

<Callout variant="tip">
You may want to postpone these optional configurations until after you've worked with this feature for a while.
</Callout>

You can control the behavior of the trace observer with various configurations. Here are some of the features you can configure:

* [Configure trace observer monitoring](/docs/distributed-tracing/infinite-tracing/infinite-tracing-configure-trace-observer-monitoring/): Allows you to see the amount of traces seen and retained.
* [Configure span attribute trace filter](/docs/distributed-tracing/infinite-tracing/infinite-tracing-configure-span-attribute-trace-filter/): Make custom filter rules for spans.

> * [Infinite Tracing: Configure random trace filter](/docs/distributed-tracing/infinite-tracing/infinite-tracing-configure-random-trace-filter): Adjust the percentage of for the random trace filter.

Author: @zuluecho9, another good catch!
